cohortextractor/query_engines/mssql.py

```diff
@@ -261,15 +274,34 @@ def get_population_table_query(self, population):
 
     def get_value_expression(self, value):
         """
-        Given a single value output node, select it from its interim table
-        Return the expression to select it, and the table to select it from
+        Given a single value output node, select it from its interim table(s)
+        Return the expression to select it, and the table(s) to select it from
         """
-        if self.is_output_node(value):
+        if self.is_category_node(value):
+            value_expr, tables = self.get_case_expression(value)
+            return value_expr, tables
+        elif self.is_output_node(value):
             table = self.output_group_tables[self.get_type_and_source(value)]
             column = self.get_output_column_name(value)
             value_expr = table.c[column]
-            return value_expr, table
-        return value, None
+            return value_expr, (table,)
+        else:
+            return value, None
+
+    def get_case_expression(self, category_node):
+        category_definitions = category_node.definitions.copy()
+        statements = []
+        tables_used = set()
+        for label, category_definition in category_definitions.items():
+            source_col = category_definition.source
+            table = self.output_group_tables[self.get_type_and_source(source_col)]
+            tables_used.add(table)
+            column = table.c[source_col.column]
```

Comment:
This section here is effectively duplicating the logic in `get_value_expression` and I think we should be calling out to that instead. That means we can support arbitrary nesting of expressions.

Reply:
Sorry, not sure I understand.  This is called from `get_value_expression`, and it's only the `table=` and `column=` lines that are duplicated, and they're done for each  category definition in turn.  It's only in its own method to break up the `get_value_expression` method.  I could put it all in `get_value_expression` instead, something like this:
```
    def get_value_expression(self, value):
        tables = None
        value_expr = value
        if self.is_category_node(value):
            category_definitions = value.definitions.copy()
            statements = []
            tables = set()
            for label, category_definition in category_definitions.items():
                source_col = category_definition.source
                table = self.output_group_tables[self.get_type_and_source(source_col)]
                column = table.c[source_col.column]
                tables.add(table)
                method = getattr(column, category_definition.operator)
                statements.append((method(category_definition.value), label))
            value_expr = sqlalchemy.case(*statements, else_=value.default)            
        elif self.is_output_node(value):
            table = self.output_group_tables[self.get_type_and_source(value)]
            column = self.get_output_column_name(value)
            value_expr = table.c[column]
            tables = table,
        return value_expr, tables
```
but I'm not sure where I can easily remove duplication

Reply:
Ah right yeah, this is me jumping ahead slightly sorry. `category_definition` here is actually just a boolean value: there's nothing particularly category-specific about it. So the translation of `category_definition` into a SQLAlchemy expression is something that can be done inside `get_value_expression`. Then `get_case_expression` becomes a method that takes a mapping of boolean expressions to categories and builds a case expression out of it. What I'm trying to get at is that `get_case_expression` doesn't, and shouldn't, need to "know" anything about to handle a `category_definition`.

If that still doesn't make any sense (which wouldn't surprise me!) then feel free to merge as is and we can work on refactoring later.

Reply:
Not sure I understand still, but I've refactored a bit :shrug: